std/machines/hash/poseidon2_bb.asm

```diff
@@ -0,0 +1,288 @@
+        1109797649,
+        1389124060,
+        1378384487,
+        973925592,
+        675566589,
+        772033245,
+        402697045,
+        386924216,
+        310894738,
+        1235941928
+    ];
+
+    // The linear layer of the external round.
+    //
+    // Doesn't have to be a complete matrix multiplication, as the last round discards
+    // part of the state, so we can skip the corresponding rows in the matrix.
+    let apply_mds = |input, output_len| {
+        let dot_product = |v1, v2| array::sum(array::zip(v1, v2, |v1_i, v2_i| v1_i * v2_i));
+        array::map(array::sub_array(MDS, 0, output_len), |row| dot_product(row, input))
+    };
+
+    let external_round = constr |c_idx, input, output| {
+        // Add constants
+        let step_a = array::zip(input, EXTERNAL_ROUND_CONSTANTS[c_idx], |v, c| v + c);
+
+        // Apply S-box
+        let x3 = array::map(step_a, constr |x| { let x3; x3 = x * x * x; x3});
+        let x7 = array::zip(x3, step_a, constr |x_3, x| { let x7; x7 = x_3 * x_3 * x; x7 });
+
+        // Multiply with MDS Matrix
+        array::zip(output, apply_mds(x7, array::len(output)), |out, x| out = x);
+    };
+
+    let internal_round = constr |c_idx, input, output| {
+        // Add constant (weird, I thought the entire state was used here,
+        // but this is how Plonky3 does it).
+        let step_a = input[0] + INTERNAL_ROUND_CONSTANTS[c_idx];
+
+        // Apply S-box
+        let x3;
+        x3 = step_a * step_a * step_a;
+        let x7;
+        x7 = x3 * x3 * step_a;
+
+        // Multiply with the diffusion matrix
+        let line_sum = x7 + array::sum(array::sub_array(input, 1, STATE_SIZE - 1));
+        output[0] = (line_sum + DIFF_DIAGONAL[0] * x7) * DIFF_MULTIPLIER;
+        array::zip(
+            array::zip(
+                array::sub_array(input, 1, STATE_SIZE - 1),
+                array::sub_array(output, 1, STATE_SIZE - 1),
+                constr |in_v, out_v| (in_v, out_v)
+            ),
+            array::sub_array(DIFF_DIAGONAL, 1, STATE_SIZE - 1),
+            constr |(in_v, out_v), diag| out_v = (line_sum + diag * in_v) * DIFF_MULTIPLIER
+        );
+    };
+
+    // Creates a sequence of 4-byte sparsed addresses.
+    let address_inc = constr |addr_high, addr_low| {
+        let addr = array::zip(
+            addr_high,
+            addr_low,
+            |high, low| (high, low)
+        );
+
+        array::fold(
+            array::zip(
+                array::sub_array(addr, 0, array::len(addr) - 1),
+                array::sub_array(addr, 1, array::len(addr) - 1),
+                constr |(high, low), (next_high, next_low)| {
+                    increment_ptr(4, high, low, next_high, next_low)
+                }
+            ), [],
+            |a, b| a + b
+        )
+    };
+
+    // Calculate the addresses and load all the inputs into the first time step
+    let input_addr_high: col[STATE_SIZE];
+    let input_addr_low: col[STATE_SIZE];
```

Comment:
I think this is for another time, but I wonder if there is a way not to commit to these values. Couldn't `increment_ptr ` return expressions for the incremented address instead of constraining an witness column to be equal to it? Not sure what happens when you pass that into `increment_ptr` again though, maybe you'd get a high-degree expression?

Reply:
I am pretty sure it would extrapolate the degree limit, due to the logic to juggle the carry between the two limbs.